samples/nrf9160/serial_lte_modem/src/tcpip/th_tcpip_service.c

```diff
@@ -0,0 +1,333 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(tcpip, CONFIG_THIN_LOG_LEVEL);
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <uart.h>
+#include <string.h>
+#include <net/socket.h>
+#include "th_connect.h"
+#include "th_util.h"
+
+#define TCPIP_BUFFER_SIZE	(CONFIG_INTER_CONNECT_UART_BUF_SIZE - 4)
+#define ADDRESS_TYPE_MASK	0x80
+#define ADDRESS_LENGTH_MASK	0x7F
+
+static u8_t rx_buffer[TCPIP_BUFFER_SIZE];
+static struct tcp_client {
+	int sock; /* Socket descriptor. */
+	struct sockaddr remote; /* remote address */
+	bool connected; /* TCP connected flag */
+} client;
+
+/**@brief Handles connect request for TCP socket transport.
+ */
+int do_tcp_connect(const u8_t *param, u8_t length)
+{
+	u16_t port;
+	bool use_ip;
+	int ret;
+
+	memcpy(&port, param, 2);
+	use_ip = (*(param+2)) & ADDRESS_TYPE_MASK;
+
+	if (use_ip) {
+		ret = util_parse_host_by_ipv4(param + 3, port,
+					(void *)&client.remote);
+	} else {
+		ret = util_parse_host_by_name(param + 3, port,
+					(void *)&client.remote);
+	}
+	if (ret) {
+		LOG_ERR("Parse failed: %d", ret);
+		return ret;
+	}
+
+	client.sock = socket(client.remote.sa_family, SOCK_STREAM, IPPROTO_TCP);
+	if (client.sock < 0) {
+		LOG_ERR("socket() failed: %d", -errno);
+		return -errno;
+	}
+
+	size_t peer_addr_size = sizeof(struct sockaddr_in6);
+
+	if (client.remote.sa_family == AF_INET) {
+		peer_addr_size = sizeof(struct sockaddr_in);
+	}
+
+	ret = connect(client.sock, (struct sockaddr *)&client.remote,
+		 peer_addr_size);
+	if (ret < 0) {
+		LOG_ERR("connect() failed: %d", -errno);
+		(void)close(client.sock);
+		return -errno;
+	}
+
+	client.connected = true;
+	LOG_INF("TCP connected");
+	return 0;
+}
+
+/**@brief Handles transport disconnection requests on TCP socket transport.
+ */
+int do_tcp_disconnect(void)
+{
+	int ret;
+
+	if (!client.connected) {
+		LOG_ERR("Not connected yet");
+		return -EINVAL;
+	}
+
+	ret = close(client.sock);
+	if (ret < 0) {
+		LOG_ERR("close() failed: %d", -errno);
+		return -errno;
+	}
+
+	client.connected = false;
+	LOG_INF("TCP disconnected");
+	return 0;
+}
+
+/**@brief Handles send requests on TCP socket transport.
+ */
+int do_tcp_send(const u8_t *data, u32_t datalen)
+{
+	u32_t offset = 0;
+	int ret;
+
+	if (!client.connected) {
+		LOG_ERR("Not connected yet");
+		return -EINVAL;
+	}
+
+	LOG_HEXDUMP_DBG(data, datalen, "TX");
+
+	while (offset < datalen) {
+		ret = send(client.sock, data + offset,
+			   datalen - offset, 0);
+		if (ret < 0) {
+			LOG_ERR("send() failed: %d", -errno);
+			return -errno;
+		}
+
+		offset += ret;
+	}
+
+	LOG_INF("TCP sent");
+	return 0;
+}
+
+/**@brief Handles receive requests on TCP socket transport.
+ */
+int do_tcp_receive(void)
+{
+	int ret;
+
+	if (!client.connected) {
+		LOG_ERR("Not connected yet");
+		return -EINVAL;
+	}
+
+	while (true) {
+		ret = recv(client.sock, rx_buffer,
+			TCPIP_BUFFER_SIZE,
+			0);
+		if (ret < 0) {
```

Comment:
We should also handle a situation when peer disconnects (`ret == 0`). In that case we should no longer use the socket - it should be closed and `client.connected` set to false.

I'd actually handle receive errors (`ret < 0`) in the same way, as they typically render the socket unusable.